### PR TITLE
Fix folding transposed GPT-OSS and Llama 4 expert weights

### PR DESCRIPTION
**Pull request by MrCapricornLiu**
### What does this PR do?

Type of change: Bug fix

`mtq.fold_weight()` fails on quantized GPT-OSS and Llama 4 experts because their quantizers are named `gate_up_proj_weight_quantizer` and `down_proj_weight_quantizer`, but the parameters do not have the `_weight` suffix. The generic name lookup therefore raises an assertion.

Fold these experts through their existing calibration weight views, which also preserve the transposed orientation used by their forward pass. Reuse the existing quantizer cleanup and pass the original weight dtype so folding matches FP16/BF16 fake-quantization rounding. Other callers retain the existing FP32 default.

Materialize offloaded weights through the shared weight-access context before folding, and write the folded values back afterward. The Accelerate writeback helper also accepts the plain dictionary used by `cpu_offload`, alongside `OffloadedWeightsLoader`.

### Usage

Existing `mtq.fold_weight(model)` calls work without additional arguments. Both values of `keep_attrs` are covered.

### Testing

- On unmodified production code, all 12 new regression cases fail at the missing-weight assertion.
- 140 tests pass across `test_huggingface.py`, `test_fused_experts.py`, `test_accelerate.py`, and `test_tensor_quant_cpu.py`.
- The regression uses complete small GPT-OSS and Llama 4 models, non-square expert weights, FP32/FP16/BF16, and both attribute-retention modes. It checks model logits, parameter storage, pre-quantization scale cleanup, and repeated folding.
- The expanded regression includes 24 CPU-executed cases with real Accelerate CPU and disk offload of the routed experts. It checks full-model logits before and after folding, repeated folding, quantizer cleanup, and restoration of meta parameters. The previously published patch reproduces a meta-tensor error when folding CPU-offloaded GPT-OSS weights.
- Earlier H800 checks passed for the resident 12-case matrix with INT8 and with FP8 quantization. Those GPU checks predate this offload follow-up and have not been rerun for it.
- Changed-file pre-commit checks pass, including Ruff, mypy, licenses, and Bandit.

NVFP4, distributed models, GPU execution with offload, and full-size checkpoint performance were not tested in this follow-up. Disk-offload writeback follows the existing helper's behavior: modified tensors are retained in its CPU state dictionary; the original offload files are not rewritten.

### Before your PR is "*Ready for review*"

- Backward compatible: yes; existing generic folding retains its default dtype.
- Copied code or new PIP dependencies: none.
- Necessary tests: added to the existing Hugging Face quantization module.
- Changelog: updated.
- Upstream review: pending.

### Additional Information

Related to item 1 of #1926. This addresses the runtime failure described in #1925; it does not close the other issues listed in #1926.

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed weight-folding failures for quantized GPT-OSS and Llama 4 models with transposed expert weights.
  * Improved folding with CPU or disk offload, repeated operations, and varied data types.
  * Preserved model outputs and parameter handling during weight folding.

* **Improvements**
  * Added support for additional quantization format inputs and improved KV-cache quantization workflows.
  * Added validation for configurations with enabled weight quantization but no applicable quantizer.

* **Tests**
  * Expanded regression coverage for offload modes, repeated folding, and output consistency.

### CODE REVIEW

File: modelopt/torch/quantization/plugins/huggingface.py [hunk 1]
```diff
@@ -620,6 +620,14 @@ def iter_weights_for_calibration(self):
             weight = getattr(self, weight_name)
             yield weight.transpose(-1, -2), getattr(self, f"{weight_name}_weight_quantizer")
 
+    def fold_weight(self, keep_attrs: bool = False):
+        """Fold expert weights in the transposed orientation used by their forward."""
+        for weight, quantizer in self.iter_weights_for_calibration():
+            if isinstance(quantizer, TensorQuantizer):
+                QuantModule._fold_weight_quantizer(
+                    quantizer, (weight,), keep_attrs, quantize_dtype=weight.dtype
+                )
```
[the remark this answers is not in the document]
Comment (MrCapricornLiu) [reply modelopt/torch/quantization/plugins/huggingface.py:629]: Confirmed with real Accelerate offload: folding the published version raises a meta-tensor error. `fold_weight()` now uses the shared materialization/writeback context, and writeback accepts the dictionary backend used by `cpu_offload`. The regression covers CPU- and disk-offloaded GPT-OSS/Llama 4 experts, logits and quantizer cleanup after folding, and repeated folding; all 140 tests in the four related CPU modules pass. Disk writeback retains the existing CPU-state-dictionary behavior rather than rewriting the offload files.

File: tests/unit/torch/quantization/plugins/test_huggingface.py [hunk 2]
```diff
@@ -382,6 +382,83 @@ def __init__(self):
...
+    )
+    pairs = list(experts.iter_weights_for_calibration())
+    for weight, quantizer in pairs:
+        quantizer.pre_quant_scale = torch.linspace(0.75, 1.25, weight.shape[-1], dtype=dtype)
+    pointers = [weight.data_ptr() for weight, _ in pairs]
+
+    with torch.no_grad():
+        expected = model(input_ids).logits
+        if offload is not None:
+            accelerate = pytest.importorskip("accelerate")
```
[the remark this answers is not in the document]
Comment (MrCapricornLiu) [reply tests/unit/torch/quantization/plugins/test_huggingface.py:438]: Added the optional-dependency comment in 424c3e2b33. Changed-file pre-commit checks pass.
